astropy/time/formats.py

```diff
@@ -378,7 +378,9 @@ def _select_subfmts(cls, pattern):
         pattern='*'.  This is OK and happens when this method is used for
         validation of an out_subfmt.
         """
-        if not isinstance(pattern, str):
+        if pattern == '*':
+            return cls.subfmts
+        elif not isinstance(pattern, str):
```

Comment:
I'd prefer leaving the validation as the first step. It's not really likely but if someone puts in a bad subtype you can get unpredictable results:
```
In [5]: np.array([1,2]) == '*'                                                  
<ipython-input-5-6379f0edecc9>:1: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  np.array([1,2]) == '*'
Out[5]: False
```
I think the isinstance() is pretty fast.

Reply:
The reason I did it this way round is the other way round suffers a significant performance hit. Although a single `isinstance` call is cheap, many are not, and `Time(np.arange(t0, t0 + 1*u.year, 1*u.hour))` calls this function 87701 times.

Perhaps we can do some type checking on `pattern` before it's passed to this function instead, since it's an internal function?

Reply:
I'm not sure that the isinstance() call is a bottleneck. At least from my timing the `isinstance` should be taking about 4 ms out of 1300 ms. Am I missing something?
```
In [6]: x = '*'
In [7]: %timeit isinstance(x, str)                                                                                                             
51.9 ns ± 0.81 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [8]: 51.9e-9 * 87000                                                                                                                        
Out[8]: 0.0045153

In [9]: %time Time(np.arange(t0, t0 + 1*u.year, 1*u.hour))                                                                                     
CPU times: user 1.31 s, sys: 5.09 ms, total: 1.31 s
Wall time: 1.31 s
```

Reply:
Ah my bad - just re-run myself and you're right 👍 . I've switched the new special casing to come after the `isinstance` check.